File: libgomp/testsuite/libgomp.fortran/critical-hint-1.f90
! { dg-do compile }

subroutine example_criticial ()
  use omp_lib
  implicit none
  integer, parameter :: my_omp_hint = omp_sync_hint_contended
  integer i, a, b

  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH0) hint(my_omp_hint)  ! OK
      a = a + i;
      !$omp end critical (HASH0)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH1) hint(omp_sync_hint_none)  ! OK
      a = a + i;
      !$omp end critical (HASH1)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH2) hint(omp_sync_hint_uncontended)  ! OK
      a = a + i;
      !$omp end critical (HASH2)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH3) hint(omp_sync_hint_contended)  ! OK
      a = a + i;
      !$omp end critical (HASH3)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH4) hint(omp_sync_hint_speculative)  ! OK
      a = a + i;
      !$omp end critical (HASH4)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH5) hint(omp_sync_hint_nonspeculative)  ! OK
      a = a + i;
      !$omp end critical (HASH5)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH6) hint(omp_sync_hint_contended + omp_sync_hint_speculative)  ! OK
      a = a + i;
      !$omp end critical (HASH6)
  end do

  !$omp parallel do
  do i = 1, 10
      ! Accepted but invalid: different hint for same name.
      !$omp critical (HASH6) hint(omp_sync_hint_contended + omp_sync_hint_speculative)  ! OK
      a = a + i;
      !$omp end critical (HASH6)
  end do
  !$omp parallel do
  do i = 1, 10
      ! Accepted but invalid: Some random integer expr.
      !$omp critical (HASH) hint(1 + 2)
      a = a + i;
      !$omp end critical (HASH)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH) hint(-3)  ! { dg-error "shall be a valid constant hint expression" }
      a = a + i;
      !$omp end critical (HASH)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical (HASH2) hint(b)  ! { dg-error "shall be a valid constant hint expression" }
      a = a + i;
      !$omp end critical (HASH2)
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical () hint(omp_hint_speculative)  ! { dg-error "Invalid character in name" }
      a = a + i;
!      !$omp end critical
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical hint(omp_sync_hint_none)  ! OK
      a = a + i;
      !$omp end critical
  end do
  !$omp parallel do
  do i = 1, 10
      !$omp critical hint(omp_sync_hint_contended)  ! { dg-error "CRITICAL at .1. with HINT clause requires a NAME, except when omp_sync_hint_none is used" }
      a = a + i;
      !$omp end critical
  end do
end

File: libgomp/testsuite/libgomp.c-c++-common/critical-hint-1.c
/* { dg-do compile } */

#include <omp.h>

void
example_criticial ()
{
  int a, b;
  #pragma omp parallel for
  for (int i = 0; i < 10; ++i)
    {
      #pragma omp critical hint(omp_sync_hint_none)  /* OK */
      a += i;
      #pragma omp critical (HASH) hint(omp_sync_hint_none)  /* OK */
      a += i;
      #pragma omp critical (HASH2) hint(omp_sync_hint_uncontended)  /* OK */
      a += i;
      #pragma omp critical (HASH3) hint(omp_sync_hint_contended)  /* OK */
      a += i;
      #pragma omp critical (HASH4) hint(omp_sync_hint_speculative)  /* OK */
      a += i;
      #pragma omp critical (HASH5) hint(omp_sync_hint_nonspeculative)  /* OK */
      a += i;
      #pragma omp critical (HASH6) hint(omp_sync_hint_contended + omp_sync_hint_speculative)  /* OK */
      a += i;
      #pragma omp critical (HASH6) hint(omp_sync_hint_contended | omp_sync_hint_speculative)  /* OK */
      a += i;

      /* Accepted but invalid: different hint for same name. */
      #pragma omp critical (HASH6) hint(omp_sync_hint_uncontended + omp_sync_hint_speculative)  
      a += i;
      /* Accepted but invalid: Some random integer expr. */
      #pragma omp critical (HASH) hint(omp_sync_hint_speculative + 1 + 2)
      a += i;

      #pragma omp critical (HASH) hint(-3)  /* { dg-error "expected constant integer expression" } */
      a += i;
      #pragma omp critical (HASH2) hint(b)  /* { dg-error "constant integer expression" } */
      a += i;
/*
  Fails with gcc as 'expected identifier' and
        with g++ as "clause requires a name, except when 'omp_sync_hint_none'"
      #pragma omp critical () hint(omp_sync_hint_speculative)
      a += i;
*/
      #pragma omp critical hint(omp_sync_hint_speculative)  /* { dg-error "with 'hint' clause requires a name, except when 'omp_sync_hint_none' is used" } */
      a += i;
    }
}
